src/Features/CSharp/Portable/BraceCompletion/AbstractCurlyBraceOrBracketCompletionService.cs

```diff
@@ -0,0 +1,273 @@
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var startPoint = openingPoint;
+            var endPoint = AdjustFormattingEndPoint(text, root, startPoint, closingPoint);
+
+            var style = documentOptions.GetOption(FormattingOptions.SmartIndent);
+            if (style == FormattingOptions.IndentStyle.Smart)
+            {
+                // Set the formatting start point to be the beginning of the first word to the left 
+                // of the opening brace location.
+                // skip whitespace
+                while (startPoint >= 0 && char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+
+                // skip tokens in the first word to the left.
+                startPoint--;
+                while (startPoint >= 0 && !char.IsWhiteSpace(text[startPoint]))
+                {
+                    startPoint--;
+                }
+            }
+
+            var spanToFormat = TextSpan.FromBounds(Math.Max(startPoint, 0), endPoint);
+            var rules = document.GetFormattingRules(spanToFormat, braceFormattingIndentationRules);
+            var result = Formatter.GetFormattingResult(
+                root, SpecializedCollections.SingletonEnumerable(spanToFormat), document.Project.Solution.Workspace, documentOptions, rules, cancellationToken);
+            if (result == null)
+            {
+                return (ImmutableArray<TextChange>.Empty, closingPoint);
+            }
+
+            var newRoot = result.GetFormattedRoot(cancellationToken);
+            var newClosingPoint = newRoot.GetAnnotatedTokens(s_closingBraceSyntaxAnnotation).Single().SpanStart + 1;
+
+            var textChanges = result.GetTextChanges(cancellationToken).ToImmutableArray();
+            return (textChanges, newClosingPoint);
+
+            async Task<Document> GetDocumentWithAnnotatedClosingBraceAsync(Document document, int closingBraceEndPoint, CancellationToken cancellationToken)
+            {
+                var originalRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var closeBraceToken = originalRoot.FindToken(closingBraceEndPoint - 1);
+                Debug.Assert(IsValidClosingBraceToken(closeBraceToken));
+
+                var newCloseBraceToken = closeBraceToken.WithAdditionalAnnotations(s_closingBraceSyntaxAnnotation);
+                var root = originalRoot.ReplaceToken(closeBraceToken, newCloseBraceToken);
+                return document.WithSyntaxRoot(root);
+            }
+        }
+
+        protected virtual int AdjustFormattingEndPoint(SourceText text, SyntaxNode root, int startPoint, int endPoint)
+            => endPoint;
```

Comment:
Consider making this method abstract and add the override to `BracketBraceCompletionService`.